### PR TITLE
Fix support for dev key generation

### PR DESCRIPTION
**Pull request by Jan4V**
Fixes a bug in fusee/exo, adds new masterkey vectors, adds dev signing support into sept

### CODE REVIEW

**Comment by Jan4V** [comment]
Also added a fix for the wrong amount of fuses being reported as blown in smcGetConfig

File: exosphere/src/masterkey.c [hunk 1]
```diff
@@ -42,10 +42,9 @@ static const uint8_t mkey_vectors_dev[MASTERKEY_REVISION_MAX][0x10] =
...
     {0x37, 0xAF, 0xAB, 0x35, 0x79, 0x09, 0xD9, 0x48, 0x29, 0xD2, 0xDB, 0xA5, 0xA5, 0xF5, 0x30, 0x19}, /* Master key 06 encrypted with Master key 07. */
-    {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}, /* TODO: Master key 07 encrypted with Master key 08. */
-    {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}, /* TODO: Master key 08 encrypted with Master key 09. */
-    {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}, /* TODO: Master key 09 encrypted with Master key 0A. */
-};
+    {0xEC, 0xE1, 0x46, 0x89, 0x37, 0xFD, 0xD2, 0x15, 0x8C, 0x3F, 0x24, 0x82, 0xEF, 0x49, 0x68, 0x04}, /* Master key 07 encrypted with Master key 08. */
+    {0x43, 0x3D, 0xC5, 0x3B, 0xEF, 0x91, 0x02, 0x21, 0x61, 0x54, 0x63, 0x8A, 0x35, 0xE7, 0xCA, 0xEE}, /* Master key 08 encrypted with Master key 09. */
+    {0x6C, 0x2E, 0xCD, 0xB3, 0x34, 0x61, 0x77, 0xF5, 0xF9, 0xB1, 0xDD, 0x61, 0x98, 0x19, 0x3E, 0xD4}, /* Master key 09 encrypted with Master key 0A. */ };
```
Comment (SciresM) [inline exosphere/src/masterkey.c]: ending brace goes on its own line

File: exosphere/src/fuse.c [hunk 2]
```diff
@@ -281,8 +281,10 @@ uint32_t fuse_get_expected_fuse_version(uint32_t target_firmware) {
     if (target_firmware > ATMOSPHERE_TARGET_FIRMWARE_COUNT) {
         generic_panic();
     }
-
-    return expected_versions[target_firmware];
+    if (fuse_get_retail_type() != 0)
+        return expected_versions[target_firmware];
+    else
+        return (target_firmware > 2) ? 1 : 0;
```
Comment (SciresM) [inline exosphere/src/fuse.c]: target_firmware > N where N isn't an ATMOSPHERE_TARGET_FIRMWARE enum value is a style violation, please fix.

File: fusee/fusee-secondary/Makefile [hunk 3]
```diff
@@ -100,7 +100,7 @@ KIPFILES 	:=	loader.kip pm.kip sm.kip ams_mitm.kip spl.kip boot.kip
 BINFILES	:=	$(foreach dir,$(DATA),$(notdir $(wildcard $(dir)/*.*))) fusee-primary.bin \
 				exosphere.bin lp0fw.bin rebootstub.bin thermosphere.bin splash_screen.bmp \
 				sept-primary.bin sept-secondary_00.enc sept-secondary_01.enc emummc.kip \
-				kernel_ldr.bin $(KIPFILES)
+				sept-secondary_dev00.enc sept-secondary_dev01.enc kernel_ldr.bin $(KIPFILES)
```
Comment (SciresM) [inline fusee/fusee-secondary/Makefile]: _dev_00, not _dev00.

File: fusee/fusee-secondary/src/nxboot.c [hunk 4]
```diff
@@ -607,11 +609,21 @@ uint32_t nxboot_main(void) {
         }
 
         if (tsec_fw_size == 0x3000) {
-            sept_secondary_enc = sept_secondary_00_enc;
-            sept_secondary_enc_size = sept_secondary_00_enc_size;
+            if (!fuse_get_retail_type()) {
```
Comment (SciresM) [inline fusee/fusee-secondary/src/nxboot.c]: invert this if. The retail case should be the if, and the dev case should be the else.

File: fusee/fusee-secondary/src/nxboot.c [hunk 5]
```diff
@@ -607,11 +609,21 @@ uint32_t nxboot_main(void) {
...
+                sept_secondary_enc = sept_secondary_dev00_enc;
+                sept_secondary_enc_size = sept_secondary_dev00_enc_size;
+            } else {
+                sept_secondary_enc = sept_secondary_00_enc;
+                sept_secondary_enc_size = sept_secondary_00_enc_size;
+            }
         } else if (tsec_fw_size == 0x3300) {
-            sept_secondary_enc = sept_secondary_01_enc;
-            sept_secondary_enc_size = sept_secondary_01_enc_size;
+            if (!fuse_get_retail_type()) {
```
Comment (SciresM) [inline fusee/fusee-secondary/src/nxboot.c]: ditto @ above

File: fusee/fusee-secondary/src/nxboot.c [hunk 6]
```diff
@@ -620,12 +632,22 @@ uint32_t nxboot_main(void) {
             fatal_error("[NXBOOT] Failed to read the TSEC firmware from Package1loader!\n");
         }
         if (target_firmware >= ATMOSPHERE_TARGET_FIRMWARE_810) {
-            sept_secondary_enc = sept_secondary_01_enc;
-            sept_secondary_enc_size = sept_secondary_01_enc_size;
+            if (!fuse_get_retail_type()) {
```
Comment (SciresM) [inline fusee/fusee-secondary/src/nxboot.c]: ditto

File: fusee/fusee-secondary/src/nxboot.c [hunk 7]
```diff
@@ -620,12 +632,22 @@ uint32_t nxboot_main(void) {
...
+                sept_secondary_enc_size = sept_secondary_dev01_enc_size;
+            } else {
+                sept_secondary_enc = sept_secondary_01_enc;
+                sept_secondary_enc_size = sept_secondary_01_enc_size;
+            }
             tsec_fw_size = 0x3300;
         } else if (target_firmware >= ATMOSPHERE_TARGET_FIRMWARE_700) {
-            sept_secondary_enc = sept_secondary_00_enc;
-            sept_secondary_enc_size = sept_secondary_00_enc_size;
+            if (!fuse_get_retail_type()) {
```
Comment (SciresM) [inline fusee/fusee-secondary/src/nxboot.c]: ditto

File: sept/sept-secondary/sept_sign.py [hunk 8]
```diff
@@ -71,11 +71,12 @@ def main(argc, argv):
         code = f.read()
     if len(code) & 0xF:
         code = code + bytearray(0x10 - (len(code) & 0xF))
-    # TODO: Support dev unit crypto
     fn, fext = os.path.splitext(argv[2])
     for key in range(KEYS.NUM_KEYS):
         with open(fn + ('_%02X' % key) + fext, 'wb') as f:
             f.write(sign_encrypt_code(code, KEYS.HOVI_SIG_KEY_PRD[key], KEYS.HOVI_ENC_KEY_PRD[key], KEYS.IV[key], b'THANKS_NVIDIA_<3', key))
+        with open(fn + ('_dev%02X' % key) + fext, 'wb') as f:
```
Comment (SciresM) [inline sept/sept-secondary/sept_sign.py]: `_dev_%02X`

**Comment by SciresM** [comment]
@Jan4V I am prepared to merge this.

The septs are not exposed in the FSS0 metadata for hekate, but I think that's "fine" (I can add them separately from this PR if CTCaer wants me to -- it would require coordination, because I think "septsecondarydev00" is too big for the current string field in that header (and would cause corruption issues as such).

Do you anticipate more changes, or should I merge?

**Comment by Jan4V** [comment]
I don't think there are going to be anymore changes at this point (in this PR, anyway), as of this PR Atmosphere (with strato disabled) is able to boot both pre-6.2.0 and post-6.2.0 firmwares as of my testing, and the eventual stratosphere patches would in my opinion best fit in a new PR. In terms of the FSS0 stuff, I haven't really looked into that, because I believe currently hekate reads sept secondary from the sd card (/sept/) files anyway, but it might be something to consider exposing in the future.
So yeah, I think this is ready to be merged.

**Comment by CTCaer** [comment]
Well the dev binaries do not get reported in the FSS0, so for now it's fine:
https://github.com/Atmosphere-NX/Atmosphere/blob/471bc7cc92db0312686dff47c4b75bab40436ce9/fusee/fusee-secondary/src/start.s#L165-L187

And https://github.com/CTCaer/hekate/blob/master/bootloader/hos/fss.c#L179-L187 is for next hekate version (because people do not update the sept folder, even though you are supposed to drag n drop the whole atmosphere release. And ask why it shows failed to decrypt pkg2).

If added, I can differentiate with `sept_sec_0#` vs `sept_sec_dev_0#`.
(Naming scheme must change because of the 15 char null-terminated string constrains).

**Pull request merged**